lxd/storage/drivers/driver_ceph_volumes.go

```diff
@@ -644,7 +644,192 @@ func (d *ceph) CreateVolumeFromMigration(vol VolumeCopy, conn io.ReadWriteCloser
 
 // RefreshVolume updates an existing volume to match the state of another.
 func (d *ceph) RefreshVolume(vol VolumeCopy, srcVol VolumeCopy, refreshSnapshots []string, allowInconsistent bool, op *operations.Operation) error {
-	return genericVFSCopyVolume(d, nil, vol, srcVol, refreshSnapshots, true, allowInconsistent, op)
+	// Copy volumes with content type filesystem using the generic approach.
+	if vol.contentType == ContentTypeFS {
+		return genericVFSCopyVolume(d, nil, vol, srcVol, refreshSnapshots, true, allowInconsistent, op)
+	}
+
+	// Refresh the VMs filesystem volume too.
+	// This will recursively call this function again and fall back to the generic way of refreshing.
+	if srcVol.IsVMBlock() {
+		// Ensure that the volume's snapshots are also replaced with their filesystem counterpart.
+		var srcFsVolSnapshots []Volume
```

Comment:
just a nit but it would be nice to initialise the slices with a capacity  when you know how big they will be, e.g. `srcFsVolSnapshots := make([]Volume, 0, len(srcVol.Snapshots ))`